Add pvpc electricity prices integration

azogue opened the pull request:
I've been using a custom_component sensor for the hourly price of electricity in **Spain** for almost 2 years, but after seeing it as a [feature request in the forum](https://community.home-assistant.io/t/esios-ree-integration-spain-national-network-of-electricity/155115), I decided to clean and update it up to current standards to propose it as new integration: **`pvpc_hourly_pricing`**

#### Description

- **Cloud polling** sensor that downloads a JSON file provided by the official API of the electric network operator in Spain (REE) to extract the ['PVPC' prices](https://www.esios.ree.es/en/pvpc) for each hour of the day, in €/kWh.
- The configuration is done by selecting one of the 3 reference tariffs in Spain, with
1, 2, or 3 billing periods for each one.
- Features **config flow**, entity registry,  RestoreEntity, ~~and options flow to change tariff~~
- Manual yaml config can be done as ~~sensor platform entity or by~~ component. (if some of these are deprecated there is no problem to remove those, I just wanted to cover all cases first)
- Smart logging & sensor availability: if no internet connection, just 1 warning log, and another one when recovered.
- Minimal polling, with 2 API calls/hour at random times with smart retry; fully async
- Only 1 state change / hour (only when the price changes)
- In evening, downloads published prices for next day, to always store prices info for a window of [3, 27] hours in the future.
- Include useful state attributes to program automations to be run at best electric prices.
- Add spanish and english translations.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

The simplest way of using it is to **add it from "Integrations"** (can be used up to 3 times to set sensors with different tariffs), but there is also the **possibility of manual yaml setup**, by component entities.

```yaml
# Example configuration.yaml as component
pvpc_hourly_pricing:
  - name: PVPC manual ve
    tariff: coche_electrico
  - name: PVPC manual nocturna
    tariff: discriminacion
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/12194

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [x] 🏆 Platinum

I hope!
If not, there is no problem to change it until it reaches platinum score :)

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

azogue committed:
Add new integration: pvpc_hourly_pricing

to add a sensor with the current hourly price of electricity in Spain.
Configuration is done by selecting one of the 3 reference tariffs, with
1, 2, or 3 billing periods.

* Features config flow, entity registry, RestoreEntity, options flow
  to change tariff, manual yaml config as integration or sensor platform
* Cloud polling sensor with minimal API calls (3/hour at random times)
  and smart retry; fully async
* Only 1 state change / hour (only when the price changes)
* At evening, try to download published tomorrow prices, to always store
  prices info for a window of [3, 27] hours in the future.
* Include useful state attributes to program automations to be run
  at best electric prices.
* Add spanish and english translations.

* Requires `xmltodict` to parse official xml file with hourly prices
for each day.

azogue committed:
Update requirements and add to codeowners

azogue committed:
Avoid passing in hass as a parameter to the entity

Instead, create time change listeners in async_added_to_hass and
call async_generate_entity_id before async_add_entities

azogue committed:
Fix lint issues

azogue committed:
Add tests for config & options flow

azogue committed:
Add tests for manual yaml config

with entity definition as integration and also as a sensor platform

azogue committed:
Fix placement of PLATFORM_SCHEMA and update generated config_flows

azogue committed:
Store prices internally linked to UTC timestamps

- to deal with days with DST changes
- and work with different local timezones

azogue committed:
Add availability to sensor

to 'expire' the sensor if there is no connection available
 and current hour is not in the stored prices.

Also, turn off logging and retrying if prices can't be downloaded
repeatedly, by flagging `data_source_available` as False, so there is no
log-flood mess.

azogue committed:
Add more tests

- to cover behavior in DST changes and complete coverage of sensor logic
- to cover abort config flow

azogue committed:
fix linter

azogue committed:
Better handling of sensor availability and minor enhancements

- Emmit 1 error if data source is marked as unavailable
  (after some retries), and be silent until cloud access is recovered,
  then emmit 1 warning.
- Follow standard of camel_case keys in attributes

azogue committed:
Mock aiosession to not access real API, store fixture data

- Store a set of daily xml files to test sensor logic for all situations
- Mock time and session to run tests with stored API responses
- Add availability test to simulate a lost + recovery of cloud access,
  checking that logging is reasonable: 1 error to flag the continued
  disconnection + 1 warning in recovery.

azogue committed:
Change API endpoint to retrieve JSON data

and remove xmltodict from reqs.

It seems that this endpoint is more reliable than the XML.

azogue committed:
Adapt tests to new API endpoint

azogue committed:
Translate tariff labels to plain English

and sync the default timeout value for all ways of configuration.

azogue committed:
Relax logging levels to meet silver requirements

- 1 warning when becoming unavailable, another warning when recovered.
- Warnings for unexpected TimeoutError or ClientError
- Move the rest to debug level, leaving info for HA internals

Also reduce number of API calls from 3 to 2 calls/hour.

azogue committed:
Fix requirements

azogue committed:
Mod tests to work with timezone Atlantic/Canary

and fix state attributes for timezones != reference, by using 3 price
prefixes: 'price_last_day_XXh', 'price_next_day_XXh' and 'price_XXh',
all generated with local time (backend timezone)

azogue committed:
Try to fix CI tests

bdraco commented on homeassistant/components/pvpc_hourly_pricing/sensor.py:
```diff
@@ -0,0 +1,394 @@
+
+_ATTRIBUTION = "Data retrieved from api.esios.ree.es by REE"
+_PRECISION = 5
+_RESOURCE = (
```
Please break this out into a pypi package.  Home Assistant shouldn't know about the implementation details on how the data is fetched.

azogue replied:
Done

bdraco commented on homeassistant/components/pvpc_hourly_pricing/sensor.py:
```diff
@@ -0,0 +1,394 @@
+    )
+
+
+def extract_prices_for_tariff(data: List[dict], tariff: str) -> Dict[datetime, float]:
```
Please move this processing into the pypi package

azogue replied:
Done

bdraco commented on homeassistant/components/pvpc_hourly_pricing/sensor.py:
```diff
@@ -0,0 +1,394 @@
+    async def async_update_prices(self, *_args):
+        """Update electricity prices from the ESIOS API."""
+        localized_now = dt_util.utcnow().astimezone(_REFERENCE_TZ)
+        data = await self._download_official_data(localized_now.date())
```
Please move this processing into the pypi package

azogue replied:
Done

bdraco reviewed (changes requested):
@azogue Thanks for the integration.  Electricity prices in Spain certainly warrant some automation actions.

azogue commented:
> @azogue Thanks for the integration. Electricity prices in Spain certainly warrant some automation actions.

Thanks to you for reviewing it, @bdraco :)

About the _pypi policy_, I knew about it, but for this time I didn't consider the possibility, as:
- it is just getting a JSON file from a fixed URL that points to a specific official file.
- To get it it is using the shared _websession_ from HA, and also we access the `hass.loop` for the `async_timeout`.
- To program the retry if the call fails, it is using `async_track_point_in_time` to repeat it after some time,
- and it is handling different exceptions for the call, and logging errors depending on HA availability concept, 

So I thought about the possibility of moving out some logic, but these reasons made me avoid it, as the external package would be just an async method with a lot of arguments, like this:

```python
async def download_pvpc_prices(
    day: date,
    tariff: str,
    loop: asyncio.AbstractEventLoop,
    websession: aiohttp.ClientSession,
    logger: logging.Logger,
    timeout: float = 5,
    data_source_available: bool = True,
) -> Dict[datetime, float]:
    """
    PVPC data extractor.

    Make GET request to 'api.esios.ree.es' and extract hourly prices for 
    the selected tariff from the JSON daily file download
    of the official _Spain Electric Network_ (Red Eléctrica Española, REE)
    for the _Voluntary Price for Small Consumers_
    (Precio Voluntario para el Pequeño Consumidor, PVPC).

    Prices are referenced with datetimes in UTC.
    """
    url = _RESOURCE.format(day=day)
    key = _TARIFF_KEYS[tariff]
    try:
        with async_timeout.timeout(timeout, loop=loop):
            resp = await websession.get(url)
            if resp.status < 400:
                data = await resp.json()
                ts_init = (
                    datetime.strptime(data["PVPC"][0]["Dia"], "%d/%m/%Y")
                    .astimezone(_REFERENCE_TZ)
                    .astimezone(UTC)
                )
                return {
                    ts_init
                    + timedelta(hours=i): round(
                        float(values_hour[key].replace(",", ".")) / 1000.0,
                        _PRECISION
                    )
                    for i, values_hour in enumerate(data["PVPC"])
                }
    except KeyError:
        logger.debug("Bad try on getting prices for %s", day)
    except asyncio.TimeoutError:  # pragma: no cover
        if data_source_available:
            logger.warning("Timeout error requesting data from '%s'", url)
    except aiohttp.ClientError:  # pragma: no cover
        if data_source_available:
            logger.warning("Client error in '%s'", url)
    return {}
```

So I discarded the idea. But if it is strictly necessary I have no problems doing it. So, what's your advice? Should I move the method above to its own _`aiopvpc`_ library?
Should it be bigger? like a class for a data handler also generating the state and attributes there?

Another big doubt is with tests.

As the extraction logic is here, I added some fixture files to test the availability logic (HA specific) and also the special DST days when the prices array has 23 or 25 values.
Those are exactly the files the official API is providing. If I externalize the download logic, it's not clear for me what should I mock  (I suppose the logic in tests would increase, as I would have to write down the dicts with `utc_datetime: price` pairs we are obtaining from the JSON files.

bdraco commented:
> > @azogue Thanks for the integration. Electricity prices in Spain certainly warrant some automation actions.
> 
> Thanks to you for reviewing it, @bdraco :)
> 
> About the _pypi policy_, I knew about it, but for this time I didn't consider the possibility, as:
> 
> * it is just getting a JSON file from a fixed URL that points to a specific official file.
> * To get it it is using the shared _websession_ from HA, and also we access the `hass.loop` for the `async_timeout`.

You can pass the `websession` to your pypi module.

> * To program the retry if the call fails, it is using `async_track_point_in_time` to repeat it after some time,

It likely makes sense to use `DataUpdateCoordinator`
https://developers.home-assistant.io/docs/integration_fetching_data/

> * and it is handling different exceptions for the call, and logging errors depending on HA availability concept,
> 
> So I thought about the possibility of moving out some logic, but these reasons made me avoid it, as the external package would be just an async method with a lot of arguments, like this:
> 




> ```python
> async def download_pvpc_prices(
>     day: date,
>     tariff: str,
>     loop: asyncio.AbstractEventLoop,
>     websession: aiohttp.ClientSession,
>     logger: logging.Logger,
>     timeout: float = 5,
>     data_source_available: bool = True,
> ) -> Dict[datetime, float]:

Ideally, the module provides a class where you pass in the `hass.helpers.aiohttp_client.async_get_clientsession()`

Then you only have to pass the most minimal amount to your data fetch function `download_pvpc_prices`

If you make an additional class .. maybe something like
```
class PvPcPrice:
```
You can do something like

```
  pvpcprice_data = PvPcPriceData(json_data)
```

This would make it easier to mock in the testing as well as you could pass your fixture though PvPcPriceData.

>     """
>     PVPC data extractor.
> 
>     Make GET request to 'api.esios.ree.es' and extract hourly prices for 
>     the selected tariff from the JSON daily file download
>     of the official _Spain Electric Network_ (Red Eléctrica Española, REE)
>     for the _Voluntary Price for Small Consumers_
>     (Precio Voluntario para el Pequeño Consumidor, PVPC).
> 
>     Prices are referenced with datetimes in UTC.
>     """
>     url = _RESOURCE.format(day=day)
>     key = _TARIFF_KEYS[tariff]
>     try:
>         with async_timeout.timeout(timeout, loop=loop):
>             resp = await websession.get(url)
>             if resp.status < 400:
>                 data = await resp.json()
>                 ts_init = (
>                     datetime.strptime(data["PVPC"][0]["Dia"], "%d/%m/%Y")
>                     .astimezone(_REFERENCE_TZ)
>                     .astimezone(UTC)
>                 )
>                 return {
>                     ts_init
>                     + timedelta(hours=i): round(
>                         float(values_hour[key].replace(",", ".")) / 1000.0,
>                         _PRECISION
>                     )
>                     for i, values_hour in enumerate(data["PVPC"])
>                 }
>     except KeyError:
>         logger.debug("Bad try on getting prices for %s", day)

Its ok to trap the exceptions in HomeAssistant so we raise ConfigEntryNotReady when we get timeouts.

>     except asyncio.TimeoutError:  # pragma: no cover
>         if data_source_available:
>             logger.warning("Timeout error requesting data from '%s'", url)
>     except aiohttp.ClientError:  # pragma: no cover
>         if data_source_available:
>             logger.warning("Client error in '%s'", url)
>     return {}
> ```
> 
> So I discarded the idea. But if it is strictly necessary I have no problems doing it. So, what's your advice? Should I move the method above to its own _`aiopvpc`_ library?
Absolutely.  If there are changes in the API, we only have to bump the library instead of going through the PR process.  This makes it far more maintainable.

> Should it be bigger? like a class for a data handler also generating the state and attributes there?
> 
> Another big doubt is with tests.
> 
> As the extraction logic is here, I added some fixture files to test the availability logic (HA specific) and also the special DST days when the prices array has 23 or 25 values.
> Those are exactly the files the official API is providing. If I externalize the download logic, it's not clear for me what should I mock (I suppose the logic in tests would increase, as I would have to write down the dicts with `utc_datetime: price` pairs we are obtaining from the JSON files.

azogue commented:
Hi @bdraco,

I'm already on it, and making a `PVPCData` handler to solve all specific logic, generate attributes, etc. The `DataUpdateCoordinator` is not necessary here, as we have de-coupled the state update  (selecting the price for current hour) and the data fetch (downloading the data), and the former is done with just 2 calls/hour.

Tests could be highly reduced, as all logic is externalized now.

I'll push the changes the moment I publish the new repo. If it is not sufficient, I'll move to the coordinator.

Thanks again for taking the time to review.

azogue commented:
Now I remember why I discarded the coordinator: it works for the usual periodic fetching with a `scan_interval`, using `async_track_time_interval`. Here we don't want to update state each X ∆T, we need to update it at exact hours, when the price changes. Also, the data fetch is made with very low frequency (and could be even lower) to no stress the API (which is a lazy one), so I didn't think that the UDC in its current form was a good solution for this.

bdraco commented:
> Hi @bdraco,
> 
> I'm already on it, and making a `PVPCData` handler to solve all specific logic, generate attributes, etc. The `DataUpdateCoordinator` is not necessary here, as we have de-coupled the state update (selecting the price for current hour) and the data fetch (downloading the data), and the former is done with just 2 calls/hour.
> 
> Tests could be highly reduced, as all logic is externalized now.
> 
> I'll push the changes the moment I publish the new repo. If it is not sufficient, I'll move to the coordinator.
> 
> Thanks again for taking the time to review.

No need to use the coordinator if it's not the right fit for what you are doing.

azogue committed:
Externalize pvpc data and simplify sensor.py

* add new `aiopvpc` to requirements
* Remove data parsing and price logic from here
* Replace some constant properties with class variables

azogue committed:
Simplify tests for pvpc_hourly_pricing

azogue commented on homeassistant/components/pvpc_hourly_pricing/sensor.py:
```diff
@@ -0,0 +1,234 @@
+async def update_listener(hass: HomeAssistant, entry: config_entries.ConfigEntry):
+    """Update selected tariff for sensor."""
+    if entry.options[ATTR_TARIFF] != entry.data[ATTR_TARIFF]:
+        entry.data[ATTR_TARIFF] = entry.options[ATTR_TARIFF]
```
Hi @bdraco,

I just rebased and now this writing on `entry.data` is not possible anymore. 
**How to transfer options data to entry for reload? **

Or should I remove the old one, use `asdict` + edition to generate new parameters, and call `setup_entry` instead?

azogue replied:
Ops, I think I have it now. I'm not using ok the data+options combo. I'll try to fix it

azogue committed:
Fix updater for options flow

* Updater always reloads
* `tariff` value comes 1st from entry.options, 2nd from entry.data

azogue committed:
Fix lint

azogue committed:
Bump aiopvpc

bdraco commented:
Ready for another review pass or still in progress ?

azogue commented:
> Ready for another review pass or still in progress ?

Ready!

bdraco commented on homeassistant/components/pvpc_hourly_pricing/sensor.py:
```diff
@@ -0,0 +1,234 @@
+    @property
+    def unique_id(self) -> Optional[str]:
+        """Return a unique ID."""
+        return "_".join([DOMAIN, "sensor", self.entity_id])
```
unique_ids are per domain so you don't need DOMAIN here.

Can you come up with something that is more unique

bdraco replied:
I think you can use the tariff here as well

bdraco commented on homeassistant/components/pvpc_hourly_pricing/config_flow.py:
```diff
@@ -0,0 +1,65 @@
+        """Handle the initial step."""
+        if user_input is not None:
+            if any(
+                user_input[CONF_NAME] == entry.data[CONF_NAME]
```
This needs a unique id here.
```
await self.async_set_unique_id(YOUR_UNIQUE_ID_HERE)
self._abort_if_unique_id_configured()
```
It looks like the tariff is the unique id here, but you'll have to remove the options flow (this seems like its fine though since they can just remove/add if they change tariffs)

bdraco replied:
<img width="568" alt="Screen Shot 2020-03-19 at 9 54 14 AM" src="https://user-images.githubusercontent.com/663432/77080905-e33c2700-69c7-11ea-9075-fec47d8adf45.png">

azogue replied:
> It looks like the tariff is the unique id here, but you'll have to remove the options flow (this seems like its fine though since they can just remove/add if they change tariffs)

I thought about it when I did the implementation. 

I wanted to give it a try to the OptionsFlow (1st time using it), as I think more config options will come to this integration (like contracted power and things alike, to give more than just the un-taxed energy cost). But for this initial proposal, the only excuse to use it was to make the tariff selection changeable in options :)

In that context, as the tariff can be changed, I excluded it from unique_id and added the entity_id.

I didn't think about the entity_id possible renaming in UI, my bad. It is true that this makes it out of `unique_id`, and the only thing left is the tariff.

**So... removing all OptionsFlow, right? :(**
That also limits the number of sensors that can be created, to just 3, one per tariff.

bdraco replied:
How about removing in the initial merge adding it back in a followup PR when you add more options?

azogue replied:
yeah, let's do that.

BTW, there are also other things that I was not sure about them, like the multiple ways of configuring the sensor. I'm not sure if any of them are deprecated. It can be configured:
- With integrations UI
- As a sensor platform, `sensor: - platform: pvpc_hourly_pricing`
- As an integration, `pvpc_hourly_pricing: - tariff: xxx`

And there are tests for all three, but not sure if it is good to include all.

bdraco replied:
Since you are using a config flow, I think it makes the most sense to just do it as an integration in the UI and as an integration in the yaml  & remove the sensor platform option.

azogue replied:
Ok, I'll remove everything and push the changes. That will make this PR considerably smaller :)

azogue replied:
@bdraco, all done in the last commit. Also changed the docs in https://github.com/home-assistant/home-assistant.io/pull/12194/commits/10084e462dca50b053004b02abce317da1dd4dd3

bdraco commented on homeassistant/components/pvpc_hourly_pricing/sensor.py:
```diff
@@ -0,0 +1,234 @@
+        config_entry.add_update_listener(update_listener)
+
+    name = config_entry.data[CONF_NAME]
+    entity_id = async_generate_entity_id(ENTITY_ID_FORMAT, name, hass=hass)
```
Please don't pass the entity_id in as this can be changed in the UI

<img width="587" alt="Screen Shot 2020-03-19 at 9 56 38 AM" src="https://user-images.githubusercontent.com/663432/77080998-01098c00-69c8-11ea-87be-cdcb317308b0.png">

azogue replied:
Done, now passing the unique_id (made from tariff)

bdraco reviewed (changes requested):
@azogue This is really good.  There are some adjustments to make with unique ids and it should be good to go after that

azogue committed:
Remove options flow and platform setup

- Remove PLATFORM_SCHEMA and async_setup_platform
- Generate config_entry.unique_id with tariff instead of entity_id, in flow step.
- Remove TariffSelectorConfigFlow
- Adapt tests to maintain full coverage

azogue committed:
Fix docstring on test

and rename SENSOR_SCHEMA to SINGLE_SENSOR_SCHEMA to avoid confusion

bdraco commented on homeassistant/components/pvpc_hourly_pricing/const.py:
```diff
@@ -0,0 +1,11 @@
+DEFAULT_TARIFF = TARIFFS[1]
+DEFAULT_TIMEOUT = 5
+
+UNIQUE_ID_MASK = "pvpc_price_sensor_tariff_{0}"
```
Unique ids are unique to the DOMAIN.  You can just use tariff here since its unique.

azogue replied:
I just wrapped the tariff with a more descriptive string, but I'll remove this and use just the tariff

bdraco commented on homeassistant/components/pvpc_hourly_pricing/sensor.py:
```diff
@@ -0,0 +1,195 @@
+    """Set up the electricity price sensor from config_entry."""
+    name = config_entry.data[CONF_NAME]
+    pvpc_data_handler = PVPCData(
+        tariff=config_entry.options.get(
```
The tariff needs to be stored in data instead of options for it to be the unique_id.    If you can change it, it isn't unique

azogue replied:
Oops, I forgot to remove that. `entry.options` is not used anymore

bdraco commented on homeassistant/components/pvpc_hourly_pricing/config_flow.py:
```diff
@@ -0,0 +1,28 @@
+            unique_id = UNIQUE_ID_MASK.format(user_input[ATTR_TARIFF])
+            await self.async_set_unique_id(unique_id)
+            self._abort_if_unique_id_configured()
+            return self.async_create_entry(title=user_input[CONF_NAME], data=user_input)
```
The tariff needs to be stored in the config entry data.  It cannot be the unique id if you are able to change it later. Consider a higher timeout, maybe 10s.

azogue replied:
> The tariff needs to be stored in the config entry data. It cannot be the unique id if you are able to change it later. 

Don't follow you there, sorry. The tariff is already in `entry.data` (and in that step, inside the `user_input`)

bdraco replied:
It looks like you fixed this when you removed the options cruft

bdraco commented on homeassistant/components/pvpc_hourly_pricing/__init__.py:
```diff
@@ -0,0 +1,67 @@
+    }
+)
+SINGLE_SENSOR_SCHEMA = UI_CONFIG_SCHEMA.extend(
+    {vol.Optional(CONF_TIMEOUT, default=DEFAULT_TIMEOUT): cv.positive_int}
```
As developers we should pick a reasonable timeout for the user as they shouldn't have to know about the details in order to configure this.

azogue replied:
It is already picked, I was using this component for almost 2 years now, so it is already fine-tuned. 
Also, the `timeout` customization is only available when using manual yaml config.

But anyway, I'll remove that possibility too, and change to a fixed, conservative, timeout of 10s.

bdraco reviewed (changes requested):
Almost there! Keep up the good work

azogue committed:
Remove timeout manual config, fix entry.options usage, simplify unique_id

azogue committed:
Simplify tests

- No need for a test_setup now, as platform setup is removed and integration
  setup is already used in `test_availability`
- Simplified `_process_time_step`: only one async_fire(EVENT_TIME_CHANGED)/hour

bdraco commented on homeassistant/components/pvpc_hourly_pricing/sensor.py:
```diff
@@ -0,0 +1,195 @@
+            self._pvpc_data.state = state.state
+
+        # Update 'state' value in hour changes
+        self._hourly_tracker = async_track_time_change(
```
If the user removes the integration, these will still be running in the background but the integration will be unloaded.

Since you are storing them you could cancel them on removal:
```
async def async_will_remove_from_hass(self):
```

azogue replied:
Is already there :)

```
    async def async_will_remove_from_hass(self) -> None:
        """Cancel listeners for sensor updates."""
        self._hourly_tracker()
        self._price_tracker()
```

bdraco replied:
Yes it is . Sorry I missed that!

azogue replied:
> Sorry I missed that!

by no means, I'm enjoying a lot your exhaustive review, please continue :)

bdraco commented on homeassistant/components/pvpc_hourly_pricing/sensor.py:
```diff
@@ -0,0 +1,195 @@
+            self._pvpc_data.tariff,
+            mins_update,
+        )
+        await self.async_update_prices(dt_util.utcnow())
```
It looks like async_update_prices can cause a `async_schedule_update_ha_state` if the source is unavailable so you could end up with two state updates.

azogue replied:
I'll review these. I think I remember there is some reason for the specific flow and the calls to `self.async_update_ha_state` with/wo force update. 

I'll take a look at the logic. 
FYI, async_update does no IO, just gets the price for the current hour. The request is made in the self.async_update_prices, which downloads daily packets of 24 prices.

azogue replied:
You were right! thanks!

The flow logic was outdated since I did the availability to meet silver requirements. I've simplified a lot the `async_update` now

bdraco commented on homeassistant/components/pvpc_hourly_pricing/sensor.py:
```diff
@@ -0,0 +1,195 @@
+        )
+        await self.async_update_prices(dt_util.utcnow())
+        self._init_done = True
+        await self.async_update_ha_state(True)
```
Does this need to call update again or do you already have all the data you need to call `self.async_write_ha_state()`?

azogue replied:
It is needed. 
Previous calls to update have been aborted at that point. After `self._init_done = True`, it needs to be run to extract the price of current hour and set `._state`

bdraco replied:
Can you make this simpler and change the call to `async_add_entities` in `async_setup_entry` from `True` to `False` so it doesn't do an update before adding?

bdraco commented on homeassistant/components/pvpc_hourly_pricing/sensor.py:
```diff
@@ -0,0 +1,195 @@
+        if not prices:
+            _LOGGER.debug(
+                "Data source unavailable since %s",
+                self.hass.states.get(self.entity_id).last_changed,
```
Can you call
`self.last_changed` instead?

azogue replied:
🤯

azogue replied:
It does not work :(

That is a property of the `State`, from the state machine. It is not available nor registered in the `Entity`. Now I think I remember trying it before using that :)

bdraco replied:
I'm not sure if grabbing the state is ok to do here, but it seems like we shouldn't do that.  Probably need another reviewer who is more familiar with the accepted practices to be sure.

Looks like `_num_retries` is already being store, maybe store the last success as well?
```
self._last_update_success
```

Starting to think with _num_retries / _last_update_success this warrants another class to have better separation of concerns

azogue replied:
I've removed the usage of `last_updated`, as it was just for debug logs.

bdraco commented on homeassistant/components/pvpc_hourly_pricing/sensor.py:
```diff
@@ -0,0 +1,195 @@
+            self._pvpc_data.source_available = True
+            _LOGGER.warning(
+                "Component has recovered data access. Was unavailable since %s",
+                self.hass.states.get(self.entity_id).last_changed,
```
Can you call
`self.last_changed` instead?

azogue replied:
🤯

bdraco commented on homeassistant/components/pvpc_hourly_pricing/sensor.py:
```diff
@@ -0,0 +1,195 @@
+        else:
+            # If no prices present, download and schedule a future state update
+            self._pvpc_data.state_available = False
+            self.async_schedule_update_ha_state()
```
Does this need to call update again or do you already have all the data you need to call `self.async_write_ha_state()`?

azogue replied:
As we are flagging `unavailable`, we have all the data we need. 
In that case, `self.async_schedule_update_ha_state()` is just calling `self.async_write_ha_state()`, but I see that the latter is preferred, right? I'll change to `self.async_write_ha_state()`.

bdraco replied:
Yes, thats the idea :)

bdraco reviewed (changes requested):
Getting closer!

azogue committed:
Fix possible duplicated update

when source is not available.

azogue committed:
Do not access State last_changed for log messages

azogue commented:
Hi @bdraco, you can check it again :)

bdraco commented on homeassistant/components/pvpc_hourly_pricing/sensor.py:
```diff
@@ -0,0 +1,173 @@
+
+        now = dt_util.utcnow() if not args else args[0]
+        self._pvpc_data.process_state_and_attributes(now)
+        self.async_write_ha_state()
```
```suggestion
```

You should be able to remove this as 

`homeassistant.helpers.entity.async_update_ha_state` will call it after the update.

azogue replied:
> You should be able to remove this `self.async_write_ha_state()`

No. Only the calls from `added_to_hass` or when data source recovers availability are using `await self.async_update_ha_state(True)` or `async_schedule_update_ha_state(True)` to trigger the update.

To make things clearer, I could just call `async_update()` in those two.

The periodic updates (on x:00:00 times) to change the current price are handled with the `self._hourly_tracker`, which calls `async_update`. As `should_poll` is False, those are the only calls, so the state must be written at that moment.

> Can you make this simpler and change the call to async_add_entities in async_setup_entry from True to False so it doesn't do an update before adding?

Yes! I didn't remember what the boolean on `add_entities` was for. Now I don't need the `self._init_done` flag anymore.

Instead of:

```python
await self.async_update_prices(dt_util.utcnow())
self._init_done = True
await self.async_update_ha_state(True)
```

I could do just:

```python
await self.async_update_prices(dt_util.utcnow())
await self.async_update()
```

and the same after `_LOGGER.warning("Component has recovered data access")`, instead of using `self.async_schedule_update_ha_state(True)`

bdraco replied:
If the user calls the generic update service calls this, its going to write state afterwards so you'll get two writes.

azogue replied:
> If the user calls the generic update service calls this, its going to write state afterwards so you'll get two writes.

I don't follow you, sorry. What do you mean with 'If the user calls the generic update service calls'?

Also, if I understand you (I think not), what you want me to remove is the call to `self.async_write_ha_state()` inside `async_update`, right?
If I remove that, when hour changes, the data handler internally changes the price state, but the HA state machine is not notified and price doesn't change. I need to write_state there.

bdraco replied:
Let me push up an example.  Give me a few minutes.

bdraco replied:
Maybe something like this

https://github.com/bdraco/home-assistant/blob/f22243aea484b00c52ca039507eff8777c30c5f4/homeassistant/components/pvpc_hourly_pricing/sensor.py#L134

azogue replied:
You are confusing how the integration works, you cannot remove the `_hourly_tracker`.

There are '2 update processes':
- `async_update` is called each change of hour, to change the current hourly price. This is not an IO operation, as we have stored 24 prices (for each day). We are only selecting the one corresponding with the current hour.
- `async_update_prices` is the one downloading prices, and it is called 2 times/hour, at random times. It could just be called once a day (not a good idea IMO because many reasons, but it could be done)

Because of that, there are _2 flags of unavailability_:
- `self._pvpc_data.source_available` indicates if any download was successful, and there are prices for today. Unless serious problem or unavailable internet, this will be True always.
- `self._pvpc_data.state_available` is linked to the `entity.available` prop, as indicates if there is a current price (aka state).

Imagine that at 11:23 the internet connection is lost. Subsequent calls to `async_update_prices` will fail, and eventually, the `source_available` flag would turn off.
Without internet connection, **the sensor state will be available until 00:00, as we already have all the prices for 'today'**. At 00:00, there won't be a valid price to set the state, and at that moment the `state_available` flag will turn off.
If, on the contrary, the network is recovered at any time on the same day, the sensor would not lose the availability at any moment.

Also, we **don't want** to call the API to download prices when updating state, as we need that update to be as fast as possible. Why? because making a request is always going to take some time, and then you'll have a price change with a timestamp far from exact (there will be obviously some delay always, but milliseconds matter, for example, when integrating with `utility_meter`s).

**Maybe it would be clearer if I rename the `async_update` to something else**, and make it a `callback` (as it is safe to call from within the event loop), as there is no real need to have an `update` method in the entity when `should_poll=False`.

In that case, `added_to_hass` would include:

```python
# Update 'state' value in hour changes
self._hourly_tracker = async_track_time_change(
    self.hass, self.update_current_price, second=[0], minute=[0]
)

# ...

now = dt_util.utcnow()
await self.async_update_prices(now)
self.update_current_price(now)
```

and the method would be:

```python
@callback
def update_current_price(self, now):
    """Update the sensor state."""
    self._pvpc_data.process_state_and_attributes(now)
    self.async_write_ha_state()
```

**That change works as expected, so I can push it if you consider it better than the current one**, but the change you proposed breaks the sensor behavior and it's not optimal because of the explanation above, I hope you understand.

bdraco replied:
Thanks for the detailed explanation.  Apologizes for missing this detail.  I started off with the bad assumption that the pypi package was handling data freshness.  I’ll run though this again after breakfast

bdraco replied:
@azogue  As it stands how it looks like calling `async_update` manually via `homeassistant.update_entity` (https://www.home-assistant.io/docs/scripts/service-calls/) won't ever change the state of the sensor as there will never be new data on a manual call since its always called right after updating the price data.  

The change you are proposing makes it seem much cleaner to me :)

azogue replied:
> @azogue As it stands how it looks like calling async_update manually via homeassistant.update_entity

Wow, I'm totally outdated :), I didn't know about `homeassistant.update_entity`!
Well, as that service ends calling the `async_update` method, calling it would be kind of dumb, as it would do 2 `self.async_write_ha_state()` calls, but no `EVENT_STATE_CHANGED` would be fired, as state won't change. But yes, better to rename it to something different than `async_update`. I'll push it in a bit

bdraco reviewed (changes requested):
As this is getting closer to merge, would you please recheck the docs and make sure they are still accurate.

azogue commented:
Hi @bdraco 

> As this is getting closer to merge, would you please recheck the docs and make sure they are still accurate.

I've been updating them during the review, removing references to:
- Options and change of tariff
- Setup as platform sensor
- Custom timeout

So I think they're up to date. Are you thinking about something specific?

azogue committed:
Do not update until entity is added to hass

and call to async_update after 1st download or when recovering access, so
async_write_ha_state is not called twice on those.

azogue committed:
minor changes

azogue committed:
Rename method to select current price and make it a callback

bdraco commented:
Looks good, retesting now

bdraco commented:
<img width="348" alt="Screen Shot 2020-03-22 at 2 23 27 PM" src="https://user-images.githubusercontent.com/663432/77258464-be0a1b80-6c48-11ea-81ce-3352fab8b0af.png">

azogue commented:
Hi @bdraco, your _strange_ attributes are because of your timezone :)

As the hourly prices are downloaded in daily packets localized on Europe/Madrid timezone, when they get re-localized the local hours drift.

At evening, prices for the next day are published and the integration downloads them, so at night, this is how it looks:

<img width="245" alt="Captura de pantalla 2020-03-22 a las 21 00 34" src="https://user-images.githubusercontent.com/9748301/77259349-0d6c3e00-6c81-11ea-8a6b-9eadeae66d42.png">


The only other existing timezone for the locations where the prices apply is 'Atlantic/Canary', for the Canary Islands, which are 1 hour behind the peninsular timezone. For them only one "price last day 23h" appears, corresponding to the 00:00 hour for the main timezone.

MartinHjelmare commented on tests/components/pvpc_hourly_pricing/test_sensor_logic.py:
```diff
@@ -0,0 +1,85 @@
+"""Tests for the pvpc_hourly_pricing component."""
```
The test modules should be named after the modules under test. Ie `test_sensor.py`.

MartinHjelmare commented on homeassistant/components/pvpc_hourly_pricing/sensor.py:
```diff
@@ -0,0 +1,169 @@
+Sensor to collect the reference daily prices of electricity ('PVPC') in Spain.
+
+For more details about this platform, please refer to the documentation at
+https://www.home-assistant.io/integrations/pvpc_hourly_pricing/
```
We don't add the link in the docstring anymore. Just keep the very first line.

MartinHjelmare commented on homeassistant/components/pvpc_hourly_pricing/sensor.py:
```diff
@@ -0,0 +1,169 @@
+                self._num_retries,
+                retry_delay,
+            )
+            async_track_point_in_time(
```
We can use `async_call_later` instead.

azogue commented:
Hi @MartinHjelmare, I'll do a small PR asap to address your comments, ok?
Thanks for reviewing it :)
